### PR TITLE
Control Task

### PR DESCRIPTION
**Pull request by davidarps**
Implementation of the sequence labeling control task described by [Hewitt and Liang (2019)](https://aclanthology.org/D19-1275.pdf).
The control task is used to assess the selectivity of a probe. Selectivity is the performance difference between the probe and performance on a control task with random, word-based labels. It quantifies how well a probe memorizes the task based on word types. 

I did not render the docs yet, so there could be bad formatting in the docs for `seq_labeling_ct.py`.

Changes:
- new method for creating a control task dataset in `control_task/seq_labeling_ct.py`
- tests for the method
- example usage in the example notebook

### CODE REVIEW

File: neurox/control_tasks/seq_labeling_ct.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+from collections import Counter
+
+import numpy as np
+
+
+def create_seq_labeling_control_task_datasets(
+    train_tokens,
+    sample_from="same",
+    dev_source=None,
```
Comment (fdalvi) [inline neurox/control_tasks/seq_labeling_ct.py]: Is there a reason why dev/test are optional? In all of our usecases, we have always needed all three sets for doing control task/selectivity exps?

File: neurox/control_tasks/seq_labeling_ct.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    Parameters
+    ----------
+    train_tokens : dict
+        Dictionary containing two lists of lists representing the training set, ``source`` and ``target``. As produced by :func:`dataloader. <mymodule.MyClass.foo>`
+    dev_source : list, optional
+        List containing the ``source`` tokens from the development set, as produced by ``dev_tokens['source']``
+    test_source : list, optional
+        List containing the ``source`` tokens from the test set, as produced by ``test_tokens['source']``
+    sample_from : str, optional
```
Comment (fdalvi) [inline neurox/control_tasks/seq_labeling_ct.py]: Move `sample_from` to be in the same order as function arguments (or move `sample_from` down in function arguments list

File: neurox/control_tasks/seq_labeling_ct.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        defaults to 'same'. The distribution from which control task labels are sampled.
+        'same': Labels are sampled from the same distribution as the main task labels.
+        'uniform': Labels are sampled from a uniform distribution.
+    case_sensitive: bool, optional
+        defaults to True. Sets whether the token comparison (for assigning the control task labels) is case-sensitive
+        or case-insensitive.
+
+    Returns
+    -------
+    ct_tokens : list
```
Comment (fdalvi) [inline neurox/control_tasks/seq_labeling_ct.py]: I prefer we have the abbreviation spelled out, so something like `control_task_tokens`, specifically for user facing documentation

File: tests/control_tasks/test_seq_labeling_ct.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
+import unittest
+from collections import Counter
+from cProfile import label
+from random import sample
+
+import neurox.control_tasks.seq_labeling_ct as ct
+import numpy as np
+
+# def create_seq_labeling_control_task_datasets(train_tokens, sample_from='same', dev_source=None, test_source=None):
```
Comment (fdalvi) [inline tests/control_tasks/test_seq_labeling_ct.py]: remove dead code

File: tests/control_tasks/test_seq_labeling_ct.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        )
+        the_moon = (
+            ct_tokens["source"][0][3:] == ct_tokens["source"][1][4:]
+            and ct_tokens["target"][0][3:] == ct_tokens["target"][1][4:]
+        )
+        eye_am_walking_on = (
+            ct_tokens["source"][1][0:4] == ct_tokens["source"][2][1:5]
+            and ct_tokens["target"][1][0:4] == ct_tokens["target"][2][1:5]
+        )
+        assert the_moon and eye_am_walking_on
```
Comment (fdalvi) [inline tests/control_tasks/test_seq_labeling_ct.py]: Please use the unittest module's assert functions (https://docs.python.org/3/library/unittest.html#unittest.TestCase.debug)

You can have multiple asserts, so this one can become:
```python
self.assertEqual(ct_tokens["source"][1][0:4], ct_tokens["source"][2][1:5])
self.assertEqual(ct_tokens["target"][1][0:4], ct_tokens["target"][2][1:5])
...
```

This makes it easy to figure out when and why a test case is failing (and also helps test runners have better output)

**Review by fdalvi — CHANGES_REQUESTED** [review]
Excellent work David! Apologies for getting back so late, I had to take some days off and then got bogged down with some other time-critical work.

Overall, this looks really good; I've left several comments throughout, but two overarching points I think we can discuss:

1. I think this should be part of the data module (so something like `neurox.data.control_task`). Once we do this, we can also simplify the name of the function to something like `neurox.data.control_task.create_sequence_labeling_dataset`

2. Please replace the asserts in test with `unittest` asserts (comment in that file has a link to all of them)


Finally, can you also rebase on the current master so we can squash and merge this commit and keep a clean repo history? 

Let me know if you need any help or clarification on any of the above points!

**Comment by fdalvi** [comment]
Also just quoting #15 incase you miss it there:

> For now, I have not formatted all of the repo code, because it will make updating your PRs a bit painful, so I'll do that once for the whole repo once we have your PRs in. So for the PRs themselves, just make sure the documentation looks alright and don't worry about running black or any formatting.

File: neurox/control_tasks/seq_labeling_ct.py (hunk 2 above)
Comment (davidarps) [reply neurox/control_tasks/seq_labeling_ct.py]: Done

File: neurox/control_tasks/seq_labeling_ct.py (hunk 3 above)
Comment (davidarps) [reply neurox/control_tasks/seq_labeling_ct.py]: Done

File: tests/control_tasks/test_seq_labeling_ct.py (hunk 4 above)
Comment (davidarps) [reply tests/control_tasks/test_seq_labeling_ct.py]: Done

File: tests/control_tasks/test_seq_labeling_ct.py (hunk 5 above)
Comment (davidarps) [reply tests/control_tasks/test_seq_labeling_ct.py]: Done

File: neurox/control_tasks/seq_labeling_ct.py (hunk 1 above)
Comment (davidarps) [reply neurox/control_tasks/seq_labeling_ct.py]: In some cases, I only use train and dev and no test set. For example in preliminary experiments or when working with default properties of the probe such as learning rate, number of epochs. In that case performance on dev and test is expected to be equal. You can also do that when dev and test are positional arguments. For example, the test set could be empty: 

`[train_ct_tokens, dev_ct_tokens, _] = create_seq_labeling_control_task_datasets(train_tokens, dev_tokens, [])`

I find the optional arguments more readable, but I'm fine with making both dev and test positional and adding to the docs that `[]` can be used as replacement for an empty dataset.

**Comment by davidarps** [comment]
Thank you for the very helpful feedback Fahim, and no worries about the delay! 
I addressed the requested changes in the recent commits, and made comments. 
Best, David

File: neurox/control_tasks/seq_labeling_ct.py (hunk 1 above)
Comment (fdalvi) [reply neurox/control_tasks/seq_labeling_ct.py]: I think you have a fair enough usecase, and given that, optional args is the right way to go. Thanks for the clarification.

**Comment by fdalvi** [comment]
Changes look good to me, I've merged the PR! This was a missing piece for a while, so thanks for taking the time to implementing it in the toolkit.
